### ISSUE TITLE
`redundant_allocation` should not be marked as machine applicable

### ISSUE THREAD

**Issue by jyn514** [rust-lang/rust-clippy#6243]
Same reasoning as https://github.com/rust-lang/rust-clippy/issues/6241; fixing this requires non-local changes which can't be done with span suggestions.

I tried this code (https://github.com/rust-lang/rust/blob/0da6d42f297642a60f2640ec313b879b376b9ad8/compiler/rustc_interface/src/queries.rs#L109-L111):

```rust
    fn codegen_backend(&self) -> &Lrc<Box<dyn CodegenBackend>> {
        &self.compiler.codegen_backend()
    }
```

I expected to see this happen: Clippy warns, but doesn't try to fix the suggestion.

Instead, this happened: Clippy tries to fix it, causing an error.

```
error[E0614]: type `(dyn rustc_codegen_ssa::traits::CodegenBackend + 'static)` cannot be dereferenced
   --> compiler/rustc_interface/src/queries.rs:284:43
    |
284 |                 Ok(passes::start_codegen(&***self.codegen_backend(), tcx, &*outputs.peek()))
    |                                           ^^^^^^^^^^^^^^^^^^^^^^^^^

warning: usage of `Rc<Box<T>>`
  --> compiler/rustc_interface/src/interface.rs:47:39
   |
47 |     pub fn codegen_backend(&self) -> &Lrc<Box<dyn CodegenBackend>> {
   |                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `Rc<dyn CodegenBackend>`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_allocation
```

### Meta

- `cargo clippy -V`: clippy 0.0.212 (ffa2e7a 2020-10-24)
